File: src/test/java/com/pubnub/api/endpoints/remoteaction/TestRemoteAction.java
package com.pubnub.api.endpoints.remoteaction;

import com.pubnub.api.PubNubError;
import com.pubnub.api.PubNubException;
import com.pubnub.api.callbacks.PNCallback;
import com.pubnub.api.models.consumer.PNStatus;
import org.jetbrains.annotations.NotNull;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;

public class TestRemoteAction<Output> implements RemoteAction<Output> {

    private final Output output;
    private final FailingStrategy failingStrategy;
    private final Executor executor = Executors.newSingleThreadExecutor();
    private final AtomicInteger asyncCallmeter = new AtomicInteger(0);
    private final AtomicInteger callsToFail;
    private PNCallback<Output> callback;

    TestRemoteAction(Output output, FailingStrategy failingStrategy) {
        this.output = output;
        this.failingStrategy = failingStrategy;
        this.callsToFail = new AtomicInteger(failingStrategy.numberOfCalls);
    }

    public static <T> TestRemoteAction<T> failing() {
        return new TestRemoteAction<>(null, FailingStrategy.ALWAYS_FAIL);
    }

    public static <T> TestRemoteAction<T> failingFirstCall(T output) {
        return new TestRemoteAction<>(output, FailingStrategy.FAIL_FIRST_CALLS);
    }

    public static <T> TestRemoteAction<T> successful(T output) {
        return new TestRemoteAction<>(output, FailingStrategy.NEVER_FAIL);
    }

    @Override
    public Output sync() throws PubNubException {
        if (failingStrategy == FailingStrategy.ALWAYS_FAIL) {
            throw PubNubException.builder().pubnubError(PubNubError.builder().errorCode(500).build()).build();
        } else if (failingStrategy == FailingStrategy.FAIL_FIRST_CALLS && this.callsToFail.getAndDecrement() > 0) {
            throw PubNubException.builder().pubnubError(PubNubError.builder().errorCode(500).build()).build();
        } else {
            return output;
        }
    }

    @Override
    public void async(@NotNull PNCallback<Output> callback) {
        this.callback = callback;
        asyncCallmeter.incrementAndGet();
        executor.execute(() -> {
            if (failingStrategy == FailingStrategy.ALWAYS_FAIL) {
                callback.onResponse(null, PNStatus.builder().error(true).build());
            } else if (failingStrategy == FailingStrategy.FAIL_FIRST_CALLS && this.callsToFail.getAndDecrement() > 0) {
                callback.onResponse(null, PNStatus.builder().error(true).build());
            } else {
                callback.onResponse(output, PNStatus.builder().build());
            }
        });
    }

    @Override
    public void retry() {
        if (callback != null) {
            async(this.callback);
        }
    }

    @Override
    public void silentCancel() {

    }

    public int howManyTimesAsyncCalled() {
        return asyncCallmeter.get();
    }

    enum FailingStrategy {
        NEVER_FAIL(0),
        ALWAYS_FAIL(0),
        FAIL_FIRST_CALLS(1);

        int numberOfCalls;

        FailingStrategy(int numberOfCalls) {
            this.numberOfCalls = numberOfCalls;
        }
    }
}

File: kotlin/src/test/kotlin/com/pubnub/api/legacy/endpoints/remoteaction/TestRemoteAction.kt
package com.pubnub.api.legacy.endpoints.remoteaction

import com.pubnub.api.PubNubError
import com.pubnub.api.PubNubException
import com.pubnub.api.endpoints.remoteaction.ExtendedRemoteAction
import com.pubnub.api.enums.PNOperationType
import com.pubnub.api.enums.PNStatusCategory
import com.pubnub.api.models.consumer.PNStatus
import java.util.concurrent.Executor
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicInteger

class TestRemoteAction<Output> internal constructor(
    private val output: Output?,
    private val failingStrategy: FailingStrategy
) : ExtendedRemoteAction<Output> {
    private val executor: Executor = Executors.newSingleThreadExecutor()
    private val asyncCallmeter = AtomicInteger(0)
    private val callsToFail: AtomicInteger
    private lateinit var callback: (result: Output?, status: PNStatus) -> Unit

    @Throws(PubNubException::class)
    override fun sync(): Output {
        return if (failingStrategy == FailingStrategy.ALWAYS_FAIL) {
            throw PubNubException(PubNubError.INTERNAL_ERROR)
        } else if (failingStrategy == FailingStrategy.FAIL_FIRST_CALLS && callsToFail.getAndDecrement() > 0) {
            throw PubNubException(PubNubError.INTERNAL_ERROR)
        } else {
            output!!
        }
    }

    override fun async(callback: (result: Output?, status: PNStatus) -> Unit) {
        this.callback = callback
        asyncCallmeter.incrementAndGet()
        executor.execute {
            if (failingStrategy == FailingStrategy.ALWAYS_FAIL) {
                callback(
                    null,
                    PNStatus(
                        error = true,
                        operation = operationType(),
                        category = PNStatusCategory.PNBadRequestCategory
                    )
                )
            } else if (failingStrategy == FailingStrategy.FAIL_FIRST_CALLS && callsToFail.getAndDecrement() > 0) {
                callback(
                    null,
                    PNStatus(
                        error = true,
                        operation = operationType(),
                        category = PNStatusCategory.PNBadRequestCategory
                    )
                )
            } else {
                callback(
                    output,
                    PNStatus(
                        error = false,
                        operation = operationType(),
                        category = PNStatusCategory.PNAcknowledgmentCategory
                    )
                )
            }
        }
    }

    override fun retry() {
        async(callback)
    }

    override fun silentCancel() {}
    fun howManyTimesAsyncCalled(): Int {
        return asyncCallmeter.get()
    }

    internal enum class FailingStrategy(var numberOfCalls: Int) {
        NEVER_FAIL(0), ALWAYS_FAIL(0), FAIL_FIRST_CALLS(1);
    }

    companion object {
        fun <T> failing(): TestRemoteAction<T> {
            return TestRemoteAction(
                null,
                FailingStrategy.ALWAYS_FAIL
            )
        }

        fun <T> failingFirstCall(output: T): TestRemoteAction<T> {
            return TestRemoteAction(
                output,
                FailingStrategy.FAIL_FIRST_CALLS
            )
        }

        fun <T> successful(
            output: T
        ): TestRemoteAction<T> {
            return TestRemoteAction(
                output,
                FailingStrategy.NEVER_FAIL
            )
        }
    }

    init {
        callsToFail = AtomicInteger(failingStrategy.numberOfCalls)
    }

    override fun operationType(): PNOperationType {
        return PNOperationType.FileOperation
    }
}
